File: python/mxnet/optimizer/updater.py
# coding: utf-8
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Updater class."""
from __future__ import absolute_import
import pickle
import numpy
from ..base import py_str
from ..ndarray import NDArray
from ..profiler import Scope
from ..util import is_np_array
from .utils import _as_classic

__all__ = ['Updater', 'get_updater']


class Updater(object):
    """Updater for kvstore."""
    def __init__(self, optimizer):
        self.optimizer = optimizer
        self.states = {}
        self.states_synced = {}
        self.aggregate_updates = optimizer.aggregate_num > 1

    def __call__(self, index, grad, weight):
        """Updates weight given gradient and index."""
        allow_np = self.optimizer.allow_np_array if hasattr(self.optimizer, "allow_np_array") else is_np_array()
        if not isinstance(index, (list, tuple)):
            indices = [index]
            grads = [_as_classic(grad, allow_np)]
            weights = [_as_classic(weight, allow_np)]
        else:
            indices = index
            grads = _as_classic(grad, allow_np)
            weights = _as_classic(weight, allow_np)
        if weights:
            self.optimizer._set_current_context(weights[0].context.device_id)
        for i, idx in enumerate(indices):
            # convert ctypes.char_p.value back to python str if needed
            if isinstance(idx, bytes):
                indices[i] = py_str(idx)
                idx = indices[i]
            if idx not in self.states:
                with Scope("updater:optimizer_state"):
                    self.states[idx] = self.optimizer.create_state_multi_precision(idx, weights[i])
                self.states_synced[idx] = True
            elif not self.states_synced[idx]:
                self.states[idx] = \
                    self.sync_state_context(self.states[idx], weights[i].context)
                self.states_synced[idx] = True
        if self.aggregate_updates:
            # segregate values based on type
            if self.optimizer.aggregate_num is not numpy.inf:
                type_map = {}
                for i, w, g in zip(indices, weights, grads):
                    if w.dtype in type_map:
                        type_map[w.dtype].append((i, w, g))
                    else:
                        type_map[w.dtype] = [(i, w, g)]
                for idx in type_map:
                    current_index = 0
                    indices, weights, grads = zip(*type_map[idx])
                    while current_index < len(indices):
                        states = []
                        step = min(self.optimizer.aggregate_num, len(indices) - current_index)
                        for j in range(step):
                            states.append(self.states[indices[current_index + j]])
                        self.optimizer.update_multi_precision(
                            indices[current_index:current_index + self.optimizer.aggregate_num],
                            weights[current_index:current_index + self.optimizer.aggregate_num],
                            grads[current_index:current_index + self.optimizer.aggregate_num],
                            states)
                        current_index += self.optimizer.aggregate_num
            else:
                states = [self.states[i] for i in indices]
                self.optimizer.update_multi_precision(indices, weights, grads, states)
        else:
            for i, w, g in zip(indices, weights, grads):
                self.optimizer.update_multi_precision([i], [w], [g], [self.states[i]])

    def sync_state_context(self, state, context):
        """sync state context."""
        if isinstance(state, NDArray):
            return state.as_in_context(context)
        elif isinstance(state, (tuple, list)):
            synced_state = (self.sync_state_context(i, context) for i in state)
            if isinstance(state, tuple):
                return tuple(synced_state)
            else:
                return list(synced_state)
        else:
            return state

    def set_states(self, states):
        """Sets updater states."""
        states = pickle.loads(states)
        if isinstance(states, tuple) and len(states) == 2:
            self.states, self.optimizer = states
        else:
            self.states = states
        self.states_synced = dict.fromkeys(self.states.keys(), False)

    def get_states(self, dump_optimizer=False):
        """Gets updater states.

        Parameters
        ----------
        dump_optimizer : bool, default False
            Whether to also save the optimizer itself. This would also save optimizer
            information such as learning rate and weight decay schedules.
        """
        return pickle.dumps((self.states, self.optimizer) if dump_optimizer else self.states)


def get_updater(optimizer):
    """Returns a closure of the updater needed for kvstore.

    Parameters
    ----------
    optimizer: Optimizer
         The optimizer.

    Returns
    -------
    updater: function
         The closure of the updater.
    """
    return Updater(optimizer)

File: csharp-package/src/MxNet/Optimizers/Updater.cs
/*****************************************************************************
   Copyright 2018 The MxNet.Sharp Authors. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/
using MxNet.Numpy;
using System;
using System.Collections.Generic;
using System.Linq;

namespace MxNet.Optimizers
{
    public class Updater
    {
        private readonly bool aggregate_updates;
        internal Optimizer optimizer;
        internal Dictionary<int, (NDArrayDict, NDArray)> states;
        private readonly Dictionary<int, bool> states_synced;

        public Updater(Optimizer opt)
        {
            optimizer = opt;
            states = new Dictionary<int, (NDArrayDict, NDArray)>();
            states_synced = new Dictionary<int, bool>();
            aggregate_updates = opt.AggregateNum > 0;
        }

        public void Call(int index, NDArray grad, NDArray weight)
        {
            Call(new[] {index}, grad, weight);
        }

        public void Call(int[] indices, NDArrayList grads, NDArrayList weights)
        {
            if (weights != null)
                optimizer.SetCurrentContext(weights[0].ctx.GetDeviceId());

            for (var i = 0; i < indices.Length; i++)
            {
                var index = indices[i];
                if (!states.ContainsKey(index))
                {
                    states[index] = optimizer.CreateStateMultiPrecision(index, weights[i]);
                    states_synced[index] = true;
                }
                else if (!states_synced[index])
                {
                    states[i] = SyncStateContext(states[i], weights[i].ctx);
                    states_synced[index] = true;
                }
            }

            if (aggregate_updates)
            {
                var type_map = new Dictionary<string, List<(int, NDArray, NDArray)>>();
                for (var i = 0; i < indices.Length; i++)
                {
                    var w = weights[i];
                    var g = grads[i];
                    if (type_map.ContainsKey(w.dtype.Name))
                    {
                        type_map[w.dtype.Name].Add((i, w, g));
                    }
                    else
                    {
                        type_map[w.dtype.Name] = new List<(int, NDArray, NDArray)>();
                        type_map[w.dtype.Name].Add((i, w, g));
                    }
                }

                foreach (var item in type_map)
                {
                    var idx = item.Key;
                    var current_index = 0;
                    (indices, weights, grads) = (item.Value.Select(x => (x.Item1)).ToArray(), item.Value.Select(x => (x.Item2)).ToArray(),                                  item.Value.Select(x => (x.Item3)).ToArray());

                    while (current_index < indices.Length)
                    {
                        var local_states = new Dictionary<int, (NDArrayDict, ndarray)>();
                        var step = Math.Min(optimizer.AggregateNum, indices.Length - current_index);
                        
                        for (var j = 0; j < step; j++) 
                            local_states.Add(j, states[indices[current_index + j]]);

                        var forupdate = item.Value.Skip(current_index).Take(current_index + optimizer.AggregateNum).ToArray();
                        var (index, weight, grad) = (forupdate.Select(x => (x.Item1)).ToArray(), forupdate.Select(x => (x.Item2)).ToArray(), forupdate.Select(x => (x.Item3)).ToArray());
                        optimizer.UpdateMultiPrecision(index, weight, grad, local_states.Values.ToArray()); //ToDo: revisit code

                        current_index += optimizer.AggregateNum;
                    }
                }
            }
            else
            {
                for (var i = 0; i < indices.Length; i++)
                    optimizer.UpdateMultiPrecision(indices[i], weights[i], grads[i], states[i]);
            }
        }

        public (NDArrayDict, NDArray) SyncStateContext((NDArrayDict, NDArray) state, Context context)
        {
            var (dict, arr) = state;
            var dict1 = new NDArrayDict();
            foreach (var item in dict) dict1[item.Key] = item.Value.AsInContext(context);

            return (dict1, arr.AsInContext(context));
        }

        public void SetStates(string states_data)
        {
            var (states, opt) = Pickle.Loads(states_data);
            if (opt != null)
                optimizer = opt;

            this.states = states;
            states_synced.Clear();
            foreach (var item in states.Keys) states_synced.Add(item, false);
        }

        public string GetStates(bool dump_optimizer = false)
        {
            if (dump_optimizer)
                return Pickle.Dumps(states, optimizer);

            return Pickle.Dumps(states);
        }
    }
}